command/agent/sink/sink.go

```diff
@@ -0,0 +1,148 @@
+	ctx    context.Context
+	logger hclog.Logger
+	client *api.Client
+	random *rand.Rand
+}
+
+func NewSinkServer(conf *SinkServerConfig) *SinkServer {
+	ss := &SinkServer{
+		DoneCh: make(chan struct{}),
+		ctx:    conf.Context,
+		logger: conf.Logger,
+		client: conf.Client,
+		random: rand.New(rand.NewSource(int64(time.Now().Nanosecond()))),
+	}
+
+	return ss
+}
+
+// Run executes the server's run loop, which is responsible for reading
+// in new tokens and pushing them out to the various sinks.
+func (ss *SinkServer) Run(incoming chan string, sinks []Sink) {
+	if incoming == nil {
+		panic("incoming or shutdown channel are nil")
+	}
+
+	ss.logger.Info("starting sink server")
+	defer func() {
+		ss.logger.Info("sink server stopped")
+		close(ss.DoneCh)
+	}()
+
+	latestToken := new(string)
+	sinkCh := make(chan func() error, len(sinks))
+	for {
+		select {
+		case <-ss.ctx.Done():
+			return
+
+		case token := <-incoming:
+			if token != *latestToken {
+
+				// Drain the existing funcs
+			drainLoop:
+				for {
+					select {
+					case <-sinkCh:
+					default:
+						break drainLoop
+					}
+				}
+
+				*latestToken = token
+
+				for _, s := range sinks {
+					sinkFunc := func(currSink Sink, currToken string) func() error {
+						return func() error {
+							if currToken != *latestToken {
+								return nil
+							}
+							if wrapTTL := currSink.WrapTTL(); wrapTTL != 0 {
+								wrapClient, err := ss.client.Clone()
+								if err != nil {
+									return errwrap.Wrapf("error deriving client for wrapping, not writing out to sink: {{err}})", err)
+								}
+								wrapClient.SetToken(currToken)
+								wrapClient.SetWrappingLookupFunc(func(string, string) string {
+									return wrapTTL.String()
+								})
+								secret, err := wrapClient.Logical().Write("sys/wrapping/wrap", map[string]interface{}{
+									"token": currToken,
+								})
+								if err != nil {
+									return errwrap.Wrapf("error wrapping token, not writing out to sink: {{err}})", err)
+								}
+								if secret == nil {
+									return errors.New("nil secret returned, not writing out to sink")
+								}
+								if secret.WrapInfo == nil {
+									return errors.New("nil wrap info returned, not writing out to sink")
+								}
+								if m, err := jsonutil.EncodeJSON(secret.WrapInfo); err != nil {
+									return errwrap.Wrapf("error marshaling token, not writing out to sink: {{err}})", err)
+								} else {
+									currToken = string(m)
+								}
+							}
+							return currSink.WriteToken(currToken)
+						}
+					}
+					sinkCh <- sinkFunc(s, token)
+				}
+			}
+
+		case sinkFunc := <-sinkCh:
+			select {
+			case <-ss.ctx.Done():
+				return
+			default:
+			}
+
+			if err := sinkFunc(); err != nil {
+				ss.logger.Error("error returned by sink write function, retrying", "error", err)
+				backoff := 2*time.Second + time.Duration(ss.random.Int63()%int64(time.Second*2)-int64(time.Second))
+				time.AfterFunc(backoff, func() {
```

Comment:
Is the intention to block here? This will spawn a go routine and could race against the above code that empties the sinkCh

Reply:
I could change it. It won't ever be a long amount of time but still probably worth it.

Reply:
Done.